core_impl/src/main/java/com/google/instrumentation/stats/StatsSerializer.java

```diff
@@ -48,17 +59,28 @@ static void serialize(StatsContextImpl context, OutputStream output) throws IOEx
   }
 
   // Deserializes based on an serialized StatsContextProto. The encoded tags are of the form:
-  //   num_tags [key_len key_bytes value_len value_bytes]*
+  //   [tag_type key_len key_bytes value_len value_bytes]*
   static StatsContextImpl deserialize(InputStream input) throws IOException {
     try {
       StatsContextProto.StatsContext context = StatsContextProto.StatsContext.parseFrom(input);
       ByteBuffer buffer = context.getTags().asReadOnlyByteBuffer();
-      int numTags = VarInt.getVarInt(buffer);
-      HashMap<String, String> tags = new HashMap<String, String>(numTags);
-      for (int i = 0; i < numTags; i++) {
-        String key = decode(buffer);
-        String val = decode(buffer);
-        tags.put(key, val);
+      HashMap<String, String> tags = new HashMap<String, String>();
+      int limit = buffer.limit();
+      while (buffer.position() < limit) {
+        int type = VarInt.getVarInt(buffer);
+        switch (type) {
+          case VALUE_TYPE_STRING:
+            String key = decode(buffer);
+            String val = decode(buffer);
+            tags.put(key, val);
+            break;
+          case VALUE_TYPE_INTEGER:
+          case VALUE_TYPE_BOOLEAN:
+            // TODO(songya): add support for tag types integer and boolean
```

Comment:
If we run into these cases while they are unsupported, we should through an IOException (otherwise we're going to try to read the tag data as a type in the next iteration).